core/nightshade/src/nightshade.rs

```diff
@@ -0,0 +1,781 @@
+/// They are used to avoid receiving two different proposals from the same authority,
+/// and penalize such behavior.
+#[derive(Debug, Clone, Serialize, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BlockHeader {
+    /// Authority proposing the block.
+    pub author: AuthorityId,
+    /// Hash of the payload contained in the block.
+    pub hash: CryptoHash,
+}
+
+/// Triplet that describe the state of each authority in the consensus.
+///
+/// Notes:
+/// We are running consensus on authorities rather than on outcomes, `endorses` refers to an authority.
+/// "outcome" will be used instead of "authority" to avoid confusion.
+///
+/// The order of the fields are very important since lexicographical comparison is used derived from `PartialEq`.
+#[derive(Debug, Clone, Serialize, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BareState {
+    /// How much confidence we have on `endorses`.
+    primary_confidence: i64,
+    /// It is the outcome with higher confidence. (Higher `endorses` values are used as tie breaker)
+    pub endorses: BlockHeader,
+    /// Confidence of outcome with second higher confidence.
+    secondary_confidence: i64,
+}
+
+impl BareState {
+    /// Empty triplets are used as starting point believe on authorities from which
+    /// we have not received any update. This state is less than any valid triplet.
+    fn empty() -> Self {
+        Self {
+            primary_confidence: -1,
+            endorses: BlockHeader { author: 0, hash: empty_cryptohash() },
+            secondary_confidence: -1,
+        }
+    }
+
+    pub fn new(author: AuthorityId, hash: CryptoHash) -> Self {
+        Self {
+            primary_confidence: 0,
+            endorses: BlockHeader { author, hash },
+            secondary_confidence: 0,
+        }
+    }
+
+    pub fn bs_encode(&self) -> Vec<u8> {
+        self.encode().expect("Fail serializing triplet.")
+    }
+
+    pub fn sign(&self, secret_key: &BlsSecretKey) -> BlsSignature {
+        secret_key.sign(&self.bs_encode())
+    }
+
+    fn verify(&self) -> bool {
+        self.primary_confidence >= self.secondary_confidence && self.secondary_confidence >= 0
+    }
+}
+
+/// `BLSProof` contains the evidence that we can have confidence `C` on some outcome `O` (and second higher confidence is `C'`)
+/// It must have signatures from more than 2/3 authorities on triplets of the form `(C - 1, O, C')`
+///
+/// This is a lazy data structure. Aggregated signature is computed after all BLS parts are supplied.
+#[derive(Debug, Clone, Serialize)]
+struct Proof {
+    bare_state: BareState,
+    mask: Vec<bool>,
+    #[serde(with = "bs58_serializer")]
+    signature: BlsSignature,
+}
+
+impl Proof {
+    fn new(bare_state: BareState, mask: Vec<bool>, signature: BlsSignature) -> Self {
+        Self {
+            bare_state,
+            mask,
+            signature,
+        }
+    }
+
+    fn verify(&self, public_keys: &Vec<BlsPublicKey>) -> bool {
```

Comment:
Need to check presence of 2/3 of participants somewhere.

Reply:
Such a big verification miss, thanks for pointing out. The first fix was incorrect, [now should be ok](https://github.com/nearprotocol/nearcore/pull/549/commits/939467befb62094ebe7dc4ea3a264187dcc52ec6)